api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,578 @@
+    );
+  }
+
+  const generatedExam = maybeGeneratedExam.data;
+
+  if (generatedExam === null) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(`Unable to locate generated exam.`)
+    );
+  }
+
+  // Create exam attempt so, even if user disconnects, their attempt is still recorded:
+  const attempt = await mapErr(
+    this.prisma.envExamAttempt.create({
+      data: {
+        userId: user.id,
+        examId: exam.id,
+        generatedExamId: generatedExam.id,
+        startTimeInMS: Date.now(),
+        questionSets: [],
+        needsRetake: false
+      }
+    })
+  );
+
+  if (attempt.hasError) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT_CREATE_EXAM_ATTEMPT(
+        JSON.stringify(attempt.error)
+      )
+    );
+  }
+  // NOTE: Anything that goes wrong after this point needs to unwind the exam attempt.
+
+  const maybeUserExam = syncMapErr(() =>
+    constructUserExam(generatedExam, exam)
+  );
+
+  if (maybeUserExam.hasError) {
+    await this.prisma.envExamAttempt.delete({
+      where: {
+        id: attempt.data.id
+      }
+    });
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeUserExam.error))
+    );
+  }
+
+  const userExam = maybeUserExam.data;
+
+  void reply.code(200);
+  return reply.send({
+    data: {
+      exam: userExam,
+      examAttempt: attempt.data
+    }
+  });
+}
+
+/**
+ * Handles updates to an exam attempt.
+ *
+ * Requires token to be validated.
+ *
+ * TODO: Consider validating req.user.id == lastAttempt.user_id?
+ *
+ * NOTE: Currently, questions can be _unanswered_ - taken away from a previous attempt submission.
+ * Theorectically, this is fine. Practically, it is unclear when that would be useful.
+ */
+async function postExamAttemptHandler(
+  this: FastifyInstance,
+  req: UpdateReqType<typeof schemas.examEnvironmentPostExamAttempt>,
+  reply: FastifyReply
+) {
+  const { attempt } = req.body;
+
+  const { user } = req;
+  if (!user) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT('Unreachable. User not authenticated.')
+    );
+  }
+
+  const maybeAttempts = await mapErr(
+    this.prisma.envExamAttempt.findMany({
+      where: {
+        examId: attempt.examId,
+        userId: user.id
+      }
+    })
+  );
+
+  if (maybeAttempts.hasError) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeAttempts.error))
+    );
+  }
+
+  const attempts = maybeAttempts.data;
+
+  if (attempts.length === 0) {
+    void reply.code(404);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT_EXAM_ATTEMPT(
+        `No attempts found for user '${user.id}' with attempt id '${attempt.examId}'.`
+      )
+    );
+  }
+
+  const latestAttempt = attempts.reduce((latest, current) =>
+    latest.startTimeInMS > current.startTimeInMS ? latest : current
+  );
```

Comment:
I'd use `_.maxBy(attempts, 'startTimeInMS')!;`, but :shrug:

Reply:
Personally, I prefer the native code, but :bigger_shrug: